Return error if patch file passed to state file.patch is malformed

msciciel opened the pull request:
### What does this PR do?
If patch file provided for file.patch state is malformed then state
returns `Patch was already applied` but patch is not applied.
It is better to return error in such case.

### What issues does this PR fix or reference?
Fixes: #59805

### Previous Behavior
```
          ID: patch_example
    Function: file.patch
        Name: /tmp/example
      Result: True
     Comment: Patch was already applied
     Started: 12:20:50.953163
    Duration: 61.558 ms
     Changes:
```

### New Behavior
```
          ID: patch_example
    Function: file.patch
        Name: /tmp/example
      Result: False
     Comment: /usr/bin/patch: **** malformed patch at line 7:
     Started: 12:33:44.915605
    Duration: 59.202 ms
     Changes:
```

Test needs to be fixed also because it return OK(false positive) with original code but log file contains and patch is not applied:
```
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff -ur a/foo/bar/math.txt b/foo/bar/math.txt
|--- a/foo/bar/math.txt	2018-04-09 18:43:52.883205365 -0500
|+++ b/foo/bar/math.txt	2018-04-09 18:44:58.525061654 -0500
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
can't find file to patch at input line 13
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff -ur a/foo/numbers.txt b/foo/numbers.txt
|--- a/foo/numbers.txt	2018-04-09 18:43:58.014272504 -0500
|+++ b/foo/numbers.txt	2018-04-09 18:44:46.487905044 -0500
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
```

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

danielrobbins commented on salt/states/file.py:
```diff
@@ -6887,6 +6887,12 @@ def _patch(patch_file, options=None, dry_run=False):
+            return ret
+
         if pre_check["retcode"] != 0:
             # Try to reverse-apply hunks from rejects file using a dry-run.
```
Just to be super-safe, I would suggest the following change. Move your new code inside the ``if pre_check["retcode"] != 0:`` conditional, immediately after this line. Then change your conditional to be ``if not os.path.exists(patch_rejects) or os.path.getsize(patch_rejects) == 0``. If the user specifies ``reject_file`` then I do not think this file will necessarily exist. The extra ``os.path.exists()`` check will ensure this doesn't result in ``OSError``.

msciciel replied:
@danielrobbins fixed. I was trying to provide unit test but I'm lack of knowledge how to mock content of `salt://example.patch` in order to run test for `file.patch`. Please verify if current state is enough or do I need to write a proper unit test.

sagetherage commented:
@msciciel if you can get this out of a draft status it will run the tests and I can get it reviewed! Thank you!

msciciel commented:
> @msciciel if you can get this out of a draft status it will run the tests and I can get it reviewed! Thank you!

Done.

msciciel commented:
Any chance for final review and possible merge?

msciciel committed:
Return error if patch file passed to state file.patch is malformed

If patch file provided for file.patch state is malformed then state
returns `Patch was already applied` but patch is not applied.

          ID: patch_example
    Function: file.patch
        Name: /tmp/example
      Result: True
     Comment: Patch was already applied
     Started: 12:20:50.953163
    Duration: 61.558 ms
     Changes:

It is better to return error in such case.

          ID: patch_example
    Function: file.patch
        Name: /tmp/example
      Result: False
     Comment: /usr/bin/patch: **** malformed patch at line 7:
     Started: 12:33:44.915605
    Duration: 59.202 ms
     Changes:

msciciel commented:
Test need to be fixed also because it return OK but log file contains:
```
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff -ur a/foo/bar/math.txt b/foo/bar/math.txt
|--- a/foo/bar/math.txt	2018-04-09 18:43:52.883205365 -0500
|+++ b/foo/bar/math.txt	2018-04-09 18:44:58.525061654 -0500
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
can't find file to patch at input line 13
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff -ur a/foo/numbers.txt b/foo/numbers.txt
|--- a/foo/numbers.txt	2018-04-09 18:43:58.014272504 -0500
|+++ b/foo/numbers.txt	2018-04-09 18:44:46.487905044 -0500
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
```

msciciel commented:
@danielrobbins @sagetherage is there a change for a review of this fix?
